File: py_src/modules/abs_hit.py
from .hittable import Hittable, HitRecord
from .aabb import ray_itval
from functools import partial
from functools import singledispatch
from .sphere import * 
from .triangle import *
from .bvh import *
from .aabb import *

def hit(obj: Hittable):
    if str(obj).startswith("sphere"):
        return partial(hit_sphere, obj) 
    elif str(obj).startswith("Triangle"):
        return partial(hit_triangle, obj)
    elif str(obj).startswith("PBRT.bvh_node"):
        return partial(hit_bvh, obj)
    elif str(obj).startswith("PBRT.aabb"):
        return partial(hit_aabb, obj)  
    else:
        raise ValueError(f"Object {obj} is not hittable") 
    

File: cpp_src/src/abs_hit.cpp
#include "abs_hit.h"

hit_lambda hit(const Hittable& h) {
    if (dynamic_cast<const aabb*>(&h)) {
        return [](const Hittable& h, const ray_itval& rt, const HitRecord& rec) {
            return hit_aabb(dynamic_cast<const aabb&>(h), rt.r, rt.t);
        };
    } else if (dynamic_cast<const Sphere*>(&h)) {
        return [](const Hittable& h, const ray_itval& rt, const HitRecord& rec) {
            return hit_sphere(dynamic_cast<const Sphere&>(h), rt, rec);
        };
    } else if (dynamic_cast<const triangle*>(&h)) {
        return [](const Hittable& h, const ray_itval& rt, const HitRecord& rec) {
            return hit_triangle(dynamic_cast<const triangle&>(h), rt, rec);
        };
    } else if (dynamic_cast<const bvh_node*>(&h)) {
        return [](const Hittable& h, const ray_itval& rt, const HitRecord& rec) {
            return hit_bvh(dynamic_cast<const bvh_node&>(h), rt, rec);
        };
    } else {
        throw std::runtime_error("Unsupported Hittable type");
    }
}